[Experiment] Use REST API in e2e tests to build up states

kevin940726 opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description
<!-- Please describe what you have changed or added -->

Use REST API to build up/clear states in e2e tests.

### Why

Most of our e2e tests require setting up the states before actually running the tests. The way we currently do is to use puppeteer to __manually__ visit those screens and click those buttons. It works fine but it's also very time-consuming. Before each test runs, we have to visit several screens and wait for them to finish, that's dozens of unnecessary and repetitive HTTP requests. What we should do is to build up the states programmatically to skip these steps. As popularized by Cypress ([#ref-1](https://docs.cypress.io/guides/getting-started/testing-your-app#Fully-test-the-login-flow-but-only-once), [#ref-2](https://docs.cypress.io/guides/references/best-practices)), we can focus on what needs to be tested and stop wasting time on building up the states manually.

Benefits are:
- Easier to read/write tests. Since we don't have to insert 100 different blocks before we can test the interactions between them, for instance.
- Faster tests. It's quite obvious that we'll speed up tests if we skip some steps.

### Why REST API

At first, `WP-CLI` is considered and seemed like the perfect choice for the job. However, after testing it in some tasks, it seems to be too slow when running some basic tasks. For instance, running a simple `list` method to get all the saved widgets takes about 0.5 to 1 second to run. In fact, all commands take at least 0.5 to 1 second to run. I'm not sure if this is the limitation of the framework, but seems like the bottleneck here is not solvable at the time.

The next option would be to use the [REST API](https://developer.wordpress.org/rest-api/). The same task that took WP-CLI 1 second can now be done under 100ms instead. Another benefit of using REST API is that we don't need to install anything before we can use it. We can run those tests on any WordPress instance as long as they have REST API supported.

There are disadvantages too though. Some tasks cannot be done with REST API, such as activating themes and plugins. We might want to use `WP-CLI` for those tasks, or call the internal API with auth cookie instead. I'm still in between of different solutions.

### How it works

This PR tries to migrate some tasks in the widgets e2e tests to use the REST API, which includes:

1. Delete all widgets
2. Get all widgets
3. Insert a legacy widget

We can reuse the `@wordpress/api-fetch` package that we've already been using in the frontend, steadier learning curve. `api-fetch` only supports browser environments though. Fortunately, we can workaround that by patching `window.fetch` to `node-fetch`.

Some REST API require authentication too. The solution is to add a [`basic-auth`](https://github.com/WP-API/Basic-Auth) test plugin to enable [HTTP basic authentication](https://developer.mozilla.org/en-US/docs/Web/HTTP/Authentication) on every REST API endpoint. This test plugin should only be used in test environments.

Then, we wrap `apiFetch` with the authentication token in a test utility `rest`, which combines both to make the REST requests. Finally, all we left is to implement those requests in our tests. It looks something like this:

```js
import { __experimentalRest as rest } from '@wordpress/e2e-test-utils';

await rest({ path: '/wp/v2/widgets' }); // Get all the widgets.
```

There's also a `batch` util we can use to send [batched requests](https://make.wordpress.org/core/2020/11/20/rest-api-batch-framework-in-wordpress-5-6/).

```js
import { __experimentalBatch as batch } from '@wordpress/e2e-test-utils';

await batch(requests); // `requests` is an array of requests.
```

### Result

Tests | trunk | Now
--- | --- | ---
`customizing-widgets.test.js` | 153s | 122s
`editing-widgets.test.js` | 136s | 122s

The tests are being run locally on macOS. The results seem to have significant deviation between runs though.

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->
E2E Tests should pass

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->
New feature

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

kevin940726 committed:
Add basic-auth test plugin and rest-api test util

kevin940726 committed:
Refactor some widgets utils to use rest api

kevin940726 committed:
Fix php lint errors

talldan reviewed (dismissed):
I think it'll be really needed as tests start to involve much more than just posts and blocks. I know the navigation editor/block tests could definitely use this too.

Might want to gather a few more opinions/reviews before merging, but this looks good to me as a first step 👍

gziolo commented:
Have you tried creating a WP plugin that sets the initial state by executing DB calls instead of using REST API? Well, it could as well run the same REST API calls inside PHP for start.

noisysocks commented on packages/e2e-test-utils/src/widgets.js:
```diff
@@ -1,32 +1,36 @@
-	}
-
-	await deactivatePlugin( 'gutenberg-test-classic-widgets' );
+	await Promise.all(
```
Why can't we use `/batch/v1` for this as well?

kevin940726 replied:
Good question. I tried that but it seems to fail intermittently. Probably some race-condition issues going on under the hood. I didn't look into the details though.

noisysocks commented:
> Have you tried creating a WP plugin that sets the initial state by executing DB calls instead of using REST API? Well, it could as well run the same REST API calls inside PHP for start.

This is a good idea, but how would Puppeteer trigger the PHP code? It would have to make a page request (e.g. navigate to `/wp-admin?gutenerg_reset_test_state=1`) which the plugin intercepts. We'd only be saving one or two requests since the REST API supports batching.

kevin940726 commented:
Though not related, changing the implementation could probably fix #33275 🤞. If that's the case then we'll have a better understanding of the underlying issue.

kevin940726 commented:
> Have you tried creating a WP plugin that sets the initial state by executing DB calls instead of using REST API?

Hmm, I'm not sure I understand.

Are DB calls something like SQL commands? I've considered this, but the DB structure in WP is often very complicated and not recommended to be altered directly.

If it's something like calling core's PHP functions then I think that have the same effects of calling the REST API? Some core functions are not exposed to REST API though (activating themes/plugins for instance), then I think it would be a good alternative for them.

gziolo commented:
> This is a good idea, but how would Puppeteer trigger the PHP code? It would have to make a page request (e.g. navigate to /wp-admin?gutenerg_reset_test_state=1) which the plugin intercepts. We'd only be saving one or two requests since the REST API supports batching.

With WP hooks you don't need to navigate to a different page, the plugin can just run the necessary logic when loading the page where you are going to do all the testing. Another difference is that you would load the page with all REST API (or their corresponding WP function wrappers) calls executed so you would only have to run test spec because the initial conditions are already set - no need to do a page reload as well like in `Should display legacy widgets` test.

> Are DB calls something like SQL commands? I've considered this, but the DB structure in WP is often very complicated and not recommended to be altered directly.

I didn't mean literally SQL commands. Whatever is the most appropriate for your use case. If that is a single WP function to call then it's a win.

noisysocks commented:
> With WP hooks you don't need to navigate to a different page, the plugin can just run the necessary logic when loading the page where you are going to do all the testing.

Right, but how would we distinguish between a page load that happens at the beginning of a test versus one that happens mid-test?

gziolo commented:
> > With WP hooks you don't need to navigate to a different page, the plugin can just run the necessary logic when loading the page where you are going to do all the testing.
> 
> Right, but how would we distinguish between a page load that happens at the beginning of a test versus one that happens mid-test?

One way I can think of is by checking whether a special flag is set in `$_GET`. Then you can do either an HTTP redirect in PHP to an URL without the flag, or remove this flag in JavaScript to avoid executing the same logic when you refresh the page in some of the existing tests.

kevin940726 commented:
> Another difference is that you would load the page with all REST API (or their corresponding WP function wrappers) calls executed so you would only have to run test spec because the initial conditions are already set - no need to do a page reload as well like in `Should display legacy widgets` test.

The page reload is needed because the test happens after we `goto` the widgets screen. We can also call `rest()` before visiting the widgets screen to reset the states, but that would require refactoring `beforeEach`. Reloading is just a simpler method and doesn't really impact performance that much. Even if we use the WP hook approach, we would still need to reload the page in test.

kevin940726 committed:
Refactor activatePlugin/deactivatePlugin to use rest API

kevin940726 commented:
I also refactored `activatePlugin` and `deactivatePlugin` to use [REST APIs](https://make.wordpress.org/core/2020/07/16/new-and-modified-rest-api-endpoints-in-wordpress-5-5/) (thanks to @Mamaduka for the heads-up!).

kevin940726 committed:
Try to fix performance test

kevin940726 commented:
I opened #33720 as an experiment of something similar to @gziolo's approach, evaluating WordPress commands directly via REST API. Currently, there's only one use case which is to switch between themes. It feels like an overkill to try to introduce something so powerful in our testing toolchain.

I prefer using REST API to reset/build states over calling WP commands, as they are well-documented and match more closely to what users would do. We can even think of them as e2e tests for REST APIs in a way.

In dea992c, I also fixed an issue where performance tests were always using the latest `e2e-test-utils` which breaks whenever we update the package in PRs.

I think this PR is the way forward, unless there're major issues I would like to merge it and monitor it in trunk. WDYT?

gziolo commented:
> I opened #33720 as an experiment of something similar to @gziolo's approach, evaluating WordPress commands directly via REST API. Currently, there's only one use case which is to switch between themes. It feels like an overkill to try to introduce something so powerful in our testing toolchain.

I don't think it's what I had in mind, because it's using REST API to send PHP code to run through `eval`.

The alternative I considered was a plugin that is always installed in E2E tests' context and it intercepts a special param in the URL. Based on the values passed in this param, we could execute some predefined commands that have access to the full WordPress API. It would be more useful for resetting the WordPress state before all test runs, think about this code that is handled now through Puppeteer:

https://github.com/WordPress/gutenberg/blob/516819e2de3c73bcdaf8be69d886070876ab414b/packages/e2e-tests/config/setup-test-framework.js#L238-L242

kevin940726 commented:
> Based on the values passed in this param, we could execute some predefined commands that have access to the full WordPress API.

This would work too! But I would like to avoid requiring developers to know PHP while all they want to do is to write e2e tests. REST API is already an officially supported interface that we can use, and the `rest` helpers is essentially just an `apiFetch` clone too. In addition, if we can't do something via REST API, like switching themes, then mostly likely we would also want to implement that in the future too.

https://github.com/WordPress/gutenberg/blob/516819e2de3c73bcdaf8be69d886070876ab414b/packages/e2e-tests/config/setup-test-framework.js#L238-L242

I believe most of them can also be implemented using REST API, and we don't have to deprecate existing test utils to achieve that. Just like `activatePlugin/deactivatePlugin` refactor in this PR, the change could be seamless and mostly backward-compatible. The downside are that it's slightly more code and require more RTT, I don't think they are big deals though, calling several requests in sequence is still faster than manually visiting the pages.

kevin940726 committed:
Mark rest as experimental and introduce batch

kevin940726 committed:
Use batch for deleting widgets

kevin940726 committed:
Remove the classic widgets test plugin

kevin940726 commented on bin/plugin/commands/performance.js:
```diff
@@ -150,12 +150,17 @@ async function setUpGitBranch( branch, environmentDirectory ) {
+) {
 	await runShellScript(
-		`npm run test-performance -- packages/e2e-tests/specs/performance/${ testSuite }.test.js`,
+		`ENVIRONMENT_DIRECTORY=${ environmentDirectory } npm run test-performance -- packages/e2e-tests/specs/performance/${ testSuite }.test.js`,
```
Huh, maybe this won't work in Windows? 🤔

talldan replied:
Should maybe test it 😄

kevin940726 replied:
I refactored it to use the official API of Node so I think it should be fine now.

talldan commented on packages/e2e-test-utils/src/plugins.js:
```diff
@@ -0,0 +1,63 @@
+ * @param {string} str The input string,
+ * @return {string} The slug.
+ */
+function slugify( str ) {
```
Hmm, is there no other util to do this? Does `kebabCase` from lodash do the job?

kevin940726 replied:
It's not the same but I think should work in most cases so I changed to use `kebabCase` 👍

talldan commented on packages/e2e-test-utils/src/widgets.js:
```diff
@@ -1,32 +1,29 @@
-	}
-
-	await deactivatePlugin( 'gutenberg-test-classic-widgets' );
+	await batch(
```
Could this be batched with the previous batch, or are there concerns about limits? (I think the limit is 25 or something). The limit might actually be a concern here if a test ever adds lots of widgets.

kevin940726 replied:
We actually can't. I think it might have some race condition going on in those API, I didn't look into why though.

talldan replied:
Ok 👍 

Might still be worth looking into the batch limit thing (some background - https://github.com/WordPress/gutenberg/issues/28709) as a follow-up, I think this batch function should ideally split the requests array based on the limit when it's over the limit and then make sequential requests for each part.

talldan commented on packages/e2e-tests/specs/widgets/editing-widgets.test.js:
```diff
@@ -608,57 +607,34 @@ describe( 'Widgets screen', () => {
-				! spinner.classList.contains( 'is-active' )
-			);
+		// Create a search widget in the first sidebar using the default instance.
+		await rest( {
+			method: 'POST',
+			path: '/wp/v2/widgets',
+			data: {
+				id_base: 'search',
+				sidebar: 'sidebar-1',
+				instance: defaultSearchInstance,
+			},
```
I feel like over time it'd be nice to see some simple helper like `await createWidgets( widgets, sidebarId )`, but this can always evolve over time with `rest` and `batch` as the foundation.

talldan commented on packages/e2e-test-utils/src/rest-api.js:
```diff
@@ -0,0 +1,76 @@
+		headers: {
+			...( options.headers || {} ),
+			// Authorize via basic authentication and test plugin.
+			Authorization: `Basic ${ Buffer.from(
```
There's a caveat here that these utilities only work with the basic auth plugin, so it might be good to make that clear in the function docs for `rest` and `batch` (for when they're potentially non-experimental).

That dependency is an interesting one. A thought that occurs to me is how the rest utils might work in another test environment, e.g. WordPress core. I suppose the `e2e-tests` package could be imported, but feels like the plugin or its code should almost be part of the test-utils package.

kevin940726 replied:
Yeah, good point! I refactored it to use cookie/nonce to authenticate the API requests, now we don't need the plugin for `rest` and `batch` to work.

talldan replied:
That's great, nice work!

talldan reviewed (approved):
Still looks good to me, thanks for working on this.

Maybe test that it definitely works in Windows before merging.

kevin940726 committed:
Use nonce and cookie to call apis

kevin940726 committed:
Pass env vars using native method

kevin940726 committed:
Use kebabCase to replace slugify

youknowriad commented:
Is this new `ENVIRONMENT_DIRECTORY` env variable mandatory to run e2e tests and performance tests? What's its purpose?

youknowriad commented on packages/e2e-tests/jest.performance.config.js:
```diff
@@ -1,7 +1,22 @@
 	setupFiles: [ '<rootDir>/config/gutenberg-phase.js' ],
+	moduleNameMapper: {
+		// Use different versions of e2e-test-utils for different environments
+		// rather than always using the latest.
```
Why?

kevin940726 replied:
To use the PR's version of `e2e-test-utils` so that changes in that package can be picked up when running the performance tests. This is not needed anymore though since we made the changes opt-in and experimental now. Just forgot to revert it 😅.

youknowriad replied:
Unless I'm wrong, this was already the case, PRs use the PR's version of the e2e-test-utils always. There's already a `--tests-env` option in the performance PR?

youknowriad commented on packages/e2e-tests/specs/widgets/editing-widgets.test.js:
```diff
@@ -608,57 +607,34 @@ describe( 'Widgets screen', () => {
-				! spinner.classList.contains( 'is-active' )
-			);
+		// Create a search widget in the first sidebar using the default instance.
+		await rest( {
```
The tests seem less readable with all these inline rest calls. I believe you should consider hiding them being some utils as otherwise, it's not really clear what the tests are doing unless you wrote it.

kevin940726 replied:
Yep, it has been brought up here: https://github.com/WordPress/gutenberg/pull/33414#discussion_r686665673.

gziolo commented on packages/e2e-test-utils/src/rest-api.js:
```diff
@@ -0,0 +1,116 @@
+	// Only need to set them once but before any requests.
+	await Promise.all( [ setAPIRootURL, setNonce ] );
+
+	return await apiFetch( options );
```
Would it be possible to use `wp.apiFetch` instead that should have already all params set when the user is logged in. Something like this:
```js
await page.evaluate( ( params ) =>
	wp.apiFetch( params ),
	options
);
 ```

kevin940726 replied:
That would make the test depends on the page it's currently running on. Some pages don't have `wp.apiFetch` available by default, and sometimes the logged in user doesn't have the necessary permissions. It also means that a reload is required after making such requests. Calling it server-side seems like a more robust approach to me.

gziolo replied:
I must admit that it's all confusing. In the current shape, the `rest` method is doing a REST API request as an admin with hardcoded credentials to `admin` + `password`. If the goal is to make it general purpose then it's far from it. Some concerns:
- e2e tests can be executed also with sites that don't have the admin account with the default credentials
- should it be possible to use `rest` with a different user account or as logged out?
- `wp-login.php` set the cookie so it's very likely that once the test refreshes the page then the rest of the test is going to be performed as an admin

kevin940726 replied:
`rest` isn't the best name here. It's running in the nodejs process rather than the browser context, so it's independent to the test and won't affect the current logged in user. As mentioned in the description of the PR, it should only be used to setup/clear states between or during tests. I'll keep searching for a better name here, fortunately it's still being marked as experimental.
